app.go

```diff
@@ -614,9 +615,14 @@ func (a *Application) processTx(txBytes []byte, isCheckTx bool) (TxHandlerResult
 
 	r, err := a.TxHandler.ProcessTx(state, txBytes, isCheckTx)
 	if err != nil {
-		storeTx.Rollback()
-		// TODO: save receipt & hash of failed EVM tx to node-local persistent cache (not app state)
-		receiptHandler.DiscardCurrentReceipt()
+		evmRevertedError := strings.Contains(err.Error(), "evm: execution reverted")
+		if evmRevertedError && state.FeatureEnabled(EvmTxReceiptsVersion2_1Feature, false) {
+			receiptHandler.CommitCurrentReceipt()
+			storeTx.Commit()
```

Comment:
This is going to persist all state changes on error...

Reply:
I have checked, only field that it changes is account nonce

Reply:
Did you test an EVM call? What about a call to a Go contract that calls the EVM? Anyway, there's an expectation that all state changes are discarded when a tx errors out, need to be very surgical about violating that expectation. Also the nonce changes shouldn't be persisted in `CheckTx`.